eng/tools/generate-dev-builds.js

```diff
@@ -0,0 +1,241 @@
+const path = require("path");
+const jju = require("jju");
+const util = require("util");
+const fs = require("fs");
+const writeFile = util.promisify(fs.writeFile);
+const readFile = util.promisify(fs.readFile);
+
+const appendPackageData = (data, pkgSrc, pkgJson) => {
+
+  var dep = new Map();
+  for(var key in pkgJson.dependencies){
+    dep.set(key, pkgJson.dependencies[key]);
+  }
+
+  var devDep = new Map();
+  for(var key in pkgJson.devDependencies){
+    devDep.set(key, pkgJson.devDependencies[key]);
+  }
+
+  data[pkgJson.name] = {
+    src: pkgSrc,
+    ver: pkgJson.version,
+    dep: dep,
+    devDep: devDep,
+    visited: false,
+    visitedDep : false
+  };
+};
+
+const getRushPackages = async (rushPath) => {
+  const baseDir = path.dirname(rushPath);
+  const rushJson = jju.parse(await readFile(rushPath, "utf8"));
+  const packageData = {};
+
+  for (const proj of rushJson.projects) {
+    const projDir = path.join(baseDir, proj.projectFolder);
+    const packageJson = jju.parse(await readFile(path.join(projDir, "package.json"), "utf8"));
+    appendPackageData(packageData, projDir, packageJson);
+  }
+
+  return packageData;
+};
+
+function writeToPackageJson(pckJsonPath,contents){
+  console.log("Inside writeTo packageJson function for - " + pckJsonPath );
+  var data = JSON.stringify(contents,null,2);
+  try{
+      fs.access(pckJsonPath, fs.F_OK, (err) => {
+      if (err) {
+        console.error(err)
+        return
+      }
+    });
+
+    fs.writeFileSync(pckJsonPath, data);
+    console.info("File created successfully with Node.js v10 fs/promises!");
+  }
+  catch(e){
+    console.error(`\n\n${e.toString()}\n\n`);
+    process.exit(1);
+  }
+  
+}
+
+function editVersion(rushPackages, buildNo,package){
+  console.log("Inside editVersion function - "+ package);
+  var pckJsonPath = rushPackages[package].src+"/"+"package.json";
+  console.log("attempting to read..");
+  console.log(pckJsonPath);
+  var contents = JSON.parse(fs.readFileSync(pckJsonPath).toString());
+  contents.version = contents.version + "-dev-"+buildNo;
```

Comment:
Won't this blindly append "-dev-(buildId)"? So if they have a preview version they will have end up with a version like `1.0.0-preview.1-dev-(buildId)`? Is that what we want or did we want to replace preview.x with the dev build?

Based on reading the sem ver spec https://semver.org/#spec-item-11 I suspect we may want to blindly append the dev build number to ensure that it always has a higher precedence then the latest non-dev build that we have published. However this is something we probably want to have some further discussion on and update the versioning spec to match if this is the approach we want to take. @danieljurek @mitchdenny